proposals/NNNN-parameter-packs.md

```diff
@@ -0,0 +1,754 @@
+
+Otherwise, we match the common prefix and suffix as long as no pack expansion types appear on either side. After this has been done, there are three possibilities:
+
+1. Left hand side contains a single pack expansion type, right hand size contains zero or more types.
+2. Left hand side contains zero or more types, right hand side contains a single pack expansion type.
+3. Any other combination, in which case the match fails.
+
+For example:
+
+```swift
+func variadic<each T>(_: repeat each T) -> (Int, repeat each T, String) {}
+
+let fn = { x in variadic(x) as (Int, Double, Float, String) }
+```
+
+Case 3 covers the case where one of the type sequences has a pack expansion, but the other one is too short; for example, matching `(Int, repeat each T, String, Float)` against `(Int, Float)` leaves you with `(repeat each T, String)` vs `()`, which is invalid.
+
+If neither side contains a pack expansion type, Case 3 also subsumes the current behavior as implemented without this proposal, where type sequence matching always requires the two type sequences to have the same length. For example, when matching `(Int, String)` against `(Int, Float, String)`, we end up with `()` vs `(Float)`, which is invalid.
+
+The type checker derives the replacement type for `T` in the call to `variadic()` by matching the contextual return type `(Int, Double, Float, String)` against the declared return type `(Int, repeat each T, String)`. The common prefix `Int` and common suffix `String` successfully match. What remains is the pack expansion type `repeat each T` and the type sequence `Double, Float`. This successfully matches, deriving the substitution `T := {Double, Float}`.
+
+While type sequence matching is positional, the type sequences may still contain labels if we're matching two tuple types. We require the labels to match exactly when dropping the common prefix and suffix, and then we only allow Case 1 and 2 to succeed if the remaining type sequences do not contain any labels.
+
+For example, matching `(x: Int, repeat each T, z: String)` against `(x: Int, Double, y: Float, z: String)` drops the common prefix and suffix, and leaves you with the pack expansion type `repeat each T` vs the type sequence `Double, y: Float`, which fails because `Double: y: Float` contains a label.
+
+However, matching `(x: Int, repeat each T, z: String)` against `(x: Int, Double, Float, z: String)` leaves you with `repeat each T` vs `Double, Float`, which succeeds with `T := {Double, Float}`, because the labels match exactly in the common prefix and suffix, and no labels remain once we get to Case 1 above.
+
+#### Open questions
+
+It is still undecided if a substitution that would produce a one-element tuple type should instead produces the element type as a scalar, treating the tuple as if it were merely parentheses.
```

Comment:
```suggestion
It is still undecided if a substitution that would produce a one-element tuple type should instead produce the element type as a scalar, treating the tuple as if it were merely parentheses.
```

Reply:
Ah, this text needs to be updated, as we did decide to produce the element type as a scalar under substitution for single-element tuples. Thanks for reminding me, even if unintentionally 🙂

Reply:
You're welcome ^^